src/lib/dates.js

```diff
@@ -476,14 +486,14 @@ exports.formatDate = function(x, fmt, tr, formatter, calendar) {
                 cDate = Registry.getComponentMethod('calendars', 'getCal')(calendar)
                     .fromJD(dateJD);
 
-            if(tr === 'y') dateStr = yearFormatWorld(cDate);
-            else if(tr === 'm') dateStr = monthFormatWorld(cDate);
+            if(tr === 'y') dateStr = formatWorld(cDate, extraFormat.year);
+            else if(tr === 'm') dateStr = formatWorld(cDate, extraFormat.month);
             else if(tr === 'd') {
-                headStr = yearFormatWorld(cDate);
-                dateStr = dayFormatWorld(cDate);
+                headStr = formatWorld(cDate, extraFormat.year);
+                dateStr = formatWorld(cDate, extraFormat.dayMonth);
```

Comment:
`modDateFormat` calls out to `components/calendars` to get the full spectrum of [d3-to-world-cal conversions](https://github.com/plotly/plotly.js/blob/98747b0d59d04edca2cf22751de997646efdcd0d/src/components/calendars/index.js#L108-L126) - you've got the ones needed so far, but others will likely show up, and no need to reinvent the wheel with `formatWorld`.

At one point there was a performance argument for the structure we have here because we could use precompiled formatters... but we lost that benefit with the original date localization PR #2207, so now I think there would be a much more concise way to do this, something like:

```js
calendar = isWorldCalendar(calendar) && calendar;

if(!fmt) {
    if(tr === 'y') fmt = extraFormat.year;
    else if(tr === 'm') fmt = extraFormat.month;
    else if(tr === 'd') {
        fmt = extraFormat.dayMonth + '\n' + extraFormat.year;
    }
    else {
        return modDateFormat(extraFormat.dayMonthYear, x, formatter, calendar) +
            '\n' + formatTime(x, tr);
    }
}

return modDateFormat(fmt, x, formatter, calendar);
```

Reply:
Yeah, I suspected that this conversion table existed somewhere but I didn't know exactly where to look (and code search with date/format/etc returned way too many entries).

I like the new proposal, pretty clear to read.

Reply:
Just test and it does work indeed on my example (only the last `else` is incorrect, need to switch date and time).